lib/vdc/BOVCollection.cpp

```diff
@@ -0,0 +1,369 @@
+        }
+        return FOUND;
+    }
+    return NOT_FOUND;
+}
+
+// Template specialization for reading data of type std::vector<int> or std::vector<float>
+// All std::vectors returned must have a size of 3
+template<typename T> int BOVCollection::_findToken(const std::string &token, std::string &line, std::vector<T> &value, bool verbose)
+{
+    // Skip comments
+    if (line[0] == '#') { return NOT_FOUND; }
+
+    size_t pos = line.find(token);
+    if (pos != std::string::npos) {    // We found the token
+        T lineValue;
+        _findTokenValue(line);
+        std::stringstream lineStream(line);
+
+        value.clear();
+        while (lineStream >> lineValue) { value.push_back(lineValue); }
+
+        if (value.size() != 3) {
+            value.clear();
+            std::string message = token + " must be a set of three values.";
+            SetErrMsg(message.c_str());
+            return ERROR;
+        }
+
+        if (lineStream.bad()) {
+            value.clear();
+            std::cout << "FAIL v " << token << std::endl;
+            std::string message = "Invalid value for " + token + " in BOV header file.";
+            SetErrMsg(message.c_str());
+            return ERROR;
+        }
+
+        if (verbose) {
+            std::cout << std::setw(20) << token << " ";
+            for (int i = 0; i < value.size(); i++) std::cout << value[i] << " ";
+            std::cout << std::endl;
+        }
+        return FOUND;
+    }
+    return NOT_FOUND;
+}
+
+void BOVCollection::_findTokenValue(std::string &line) const
+{
+    std::string delimiter = ": ";
+
+    size_t      pos = 0;
+    std::string token;
+    while ((pos = line.find(delimiter)) != std::string::npos) {
+        token = line.substr(0, pos);
+        line.erase(0, pos + delimiter.length());
+    }
+}
+
+size_t BOVCollection::_sizeOfFormat(DC::XType type) const
+{
+    switch (type) {
+    case DC::XType::INT32: return 4;
+    case DC::XType::FLOAT: return 4;
+    case DC::XType::DOUBLE: return 8;
+    default: return -1;
+    }
+}
+
+void BOVCollection::_swapBytes(void *vptr, size_t size, size_t n) const
+{
+    unsigned char *ucptr = (unsigned char *)vptr;
+    unsigned char  uc;
+    size_t         i, j;
+
+    for (j = 0; j < n; j++) {
+        for (i = 0; i < size / 2; i++) {
+            uc = ucptr[i];
+            ucptr[i] = ucptr[size - i - 1];
+            ucptr[size - i - 1] = uc;
+        }
+        ucptr += size;
+    }
+}
+
+template<class T> int BOVCollection::ReadRegion(const std::vector<size_t> &min, const std::vector<size_t> &max, T region)
+{
+    FILE *fp = fopen(_dataFile.c_str(), "rb");
+    if (!fp) {
+        SetErrMsg("Invalid file: %d", fp);
+        return -1;
+    }
+
+    size_t formatSize = _sizeOfFormat(_dataFormat);
+    if (formatSize < 0) {
+        SetErrMsg("Unspecified data format");
+        return (-1);
+    }
+
+    if (_dataSize.size() != 3) {
+        SetErrMsg("Invalid grid size (must be 3D)");
+        return (-1);
+    }
```

Comment:
If this thing has to be size 3, why don't you declare it as a fixed sized structure, something like `std::array<float, 3>`? John has to put in a lot of effort to grid classes to deprecate variable length vectors, and I think we should avoid the same mistakes in new code.

Reply:
No problem.  Fixed.